tfhe/src/high_level_api/strings/ascii/mod.rs

```diff
@@ -0,0 +1,277 @@
+}
+
+impl EncryptableString<'_> {
+    fn str_and_padding(&self) -> (&str, Option<u32>) {
+        match self {
+            EncryptableString::NoPadding(str) => (str, None),
+            EncryptableString::WithPadding { str, padding } => (str, Some(*padding)),
+        }
+    }
+}
+
+pub(crate) enum AsciiDevice {
+    Cpu(FheString),
+}
+
+impl From<FheString> for AsciiDevice {
+    fn from(value: FheString) -> Self {
+        Self::Cpu(value)
+    }
+}
+
+impl AsciiDevice {
+    pub fn on_cpu(&self) -> MaybeCloned<'_, FheString> {
+        match self {
+            Self::Cpu(cpu_string) => MaybeCloned::Borrowed(cpu_string),
+        }
+    }
+}
+
+impl Clone for AsciiDevice {
+    fn clone(&self) -> Self {
+        match self {
+            Self::Cpu(s) => Self::Cpu(s.clone()),
+        }
+    }
+}
+
+impl serde::Serialize for AsciiDevice {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        self.on_cpu().serialize(serializer)
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for AsciiDevice {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let deserialized = Self::Cpu(FheString::deserialize(deserializer)?);
+        Ok(deserialized)
+    }
+}
+
+// Only CPU data are serialized so we only versionize the CPU type.
+#[derive(serde::Serialize, serde::Deserialize)]
+#[cfg_attr(tfhe_lints, allow(tfhe_lints::serialize_without_versionize))]
+pub(crate) struct AsciiDeviceVersionOwned(<FheString as VersionizeOwned>::VersionedOwned);
+
+#[derive(Serialize, Deserialize)]
+#[cfg_attr(tfhe_lints, allow(tfhe_lints::serialize_without_versionize))]
+pub(crate) enum AsciiDeviceVersionedOwned {
+    V0(AsciiDeviceVersionOwned),
+}
+
+impl Versionize for AsciiDevice {
+    type Versioned<'vers> = AsciiDeviceVersionedOwned;
+
+    fn versionize(&self) -> Self::Versioned<'_> {
+        let data = self.on_cpu();
+        let versioned = data.into_owned().versionize_owned();
+        AsciiDeviceVersionedOwned::V0(AsciiDeviceVersionOwned(versioned))
+    }
+}
+
+impl VersionizeOwned for AsciiDevice {
+    type VersionedOwned = AsciiDeviceVersionedOwned;
+
+    fn versionize_owned(self) -> Self::VersionedOwned {
+        let cpu_data = self.on_cpu();
+        AsciiDeviceVersionedOwned::V0(AsciiDeviceVersionOwned(
+            cpu_data.into_owned().versionize_owned(),
+        ))
+    }
+}
+
+impl Unversionize for AsciiDevice {
+    fn unversionize(versioned: Self::VersionedOwned) -> Result<Self, UnversionizeError> {
+        match versioned {
+            AsciiDeviceVersionedOwned::V0(v0) => {
+                let unversioned = Self::Cpu(FheString::unversionize(v0.0)?);
+                Ok(unversioned)
+            }
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Versionize, Clone)]
+#[versionize(FheAsciiStringVersions)]
+pub struct FheAsciiString {
+    pub(crate) inner: AsciiDevice,
```

Comment:
odd choice for the inner type name ?

Reply:
Yes, I was not sure what name to give, maybe `InnerString` might be better

Reply:
this is fully pub(crate) ?

Reply:
Yes

Reply:
ok then let's get this ready for release and update naming later